### PR TITLE
Allow JSON filtering and clarify sort order

### LINKED ISSUE #1215: Filter objects where a particular field is null
`/records?field.name=null` ?

### PR DESCRIPTION
**Pull request by glasserc**
Fixes #1215, #1216, #1257, and possibly others.

Instead of coercing all values to strings when doing comparisons in Postgres, let Postgres use its own JSONB type comparison. This gets us 90% of the way to supporting JSON comparisons, including comparisons against null. The remaining part is about handling "missing" fields. I've made the executive decision that a "missing" value is not equal to any value of any type, including the JSON null. (Fortunately, this is easy in Postgres.) We handle these explicitly by adding `field IS NOT NULL AND` or `field IS NULL OR` as needed.

Because we already support sorting on a field with mixed-type values and/or missing values, we already implicitly have an ordering defined for comparisons like LT, GT, MAX, MIN, etc. This ordering puts missing values as "bigger" than other values. While we're messing around with comparisons involving NULL, fix this too.

We don't appear to have any clear recommendation about our ordering in the memory backend. To some extent, the tests assume that the ordering is the same. However, I am skeptical that we can always harmonize on the Postgres ordering. I think a more sensible approach is to allow each backend to define its own sort order, as long as it's consistent. For this reason, I didn't struggle with making the memory backend comparison function exactly the same as the Postgres one does.

- [x] Add documentation.
- [x] Add tests.
- [x] Add a changelog entry.
- (n/a) Add your name in the contributors file.
- [x] If you changed the HTTP API, update the [API_VERSION](https://github.com/Kinto/kinto/blob/master/kinto/__init__.py#L15) constant and add an API changelog entry [in the docs](https://github.com/Kinto/kinto/blob/master/docs/api/index.rst)

r? @leplatrem, @Natim 

I'm pretty confident in the approach, but I'm not sure whether this qualifies as a major version and/or API version.

### CODE REVIEW

**Comment by glasserc** [comment]
I've added some commits to implement a `has_` operator, addressing #344 directly, since the change in how we handle missing fields breaks a test each in `test_views_groups` and `test_views_collections`. I know this is kind of outside the scope of this PR, but it touches a lot of the same code and benefits from the new cleaner implementation. If you'd rather, I can separate out the commits for that feature.

File: kinto/core/storage/memory.py [hunk 1]
```diff
@@ -332,26 +341,54 @@ def apply_filters(records, filters):
...
-                        continue
-                    else:
-                        left = ''  # To mimic what we do for postgresql.
-            matches = matches and operators[f.operator](left, right)
+                matches = matches and operators[f.operator](left, right)
         if matches:
             yield record
 
 
+def schwartz_transform(value):
```
Comment (leplatrem) [inline kinto/core/storage/memory.py]: cool, did not know about this name :) 

btw, on wikipedia it's called schwartzian transform

File: kinto/core/storage/postgresql/__init__.py [hunk 2]
```diff
@@ -686,6 +686,7 @@ def _format_conditions(self, filters, id_field, modified_field,
         holders = {}
         for i, filtr in enumerate(filters):
             value = filtr.value
+            query_is_like = filtr.operator == COMPARISON.LIKE
```
Comment (leplatrem) [inline kinto/core/storage/postgresql/__init__.py]: **µnit**:  boolean values usually start with `is_` or `has_`... what about `is_like_operator` ?

File: kinto/core/storage/postgresql/__init__.py [hunk 3]
```diff
@@ -701,43 +702,72 @@ def _format_conditions(self, filters, id_field, modified_field,
...
+                # everything, including JSON null. Do this on a
+                # per-operator basis.
+                if filtr.operator in (
+                        # NULLs aren't EQ to anything (definitionally).
+                        COMPARISON.EQ,
+                        # So they can't match anything in an INCLUDE.
+                        COMPARISON.IN,
+                        # Nor can they be LIKE anything.
+                        COMPARISON.LIKE,
+                ):
```
Comment (leplatrem) [inline kinto/core/storage/postgresql/__init__.py]: **µnit**: in python multi-lines `if` are not recommended. Would it be better to have two intermediary variables like `null_comparable_ops` and `null_incomparable_ops` ?

File: kinto/core/storage/testing.py [hunk 4]
```diff
@@ -350,6 +350,50 @@ def test_get_all_handle_sorting_on_subobject(self):
...
+    def test_get_all_sorting_is_consistent_with_filtering(self):
+        self.create_record({'flavor': 'strawberry'})
+        self.create_record({'flavor': 'blueberry', 'author': None})
+        self.create_record({'flavor': 'raspberry', 'author': 1})
+        self.create_record({'flavor': 'orange', 'author': True})
+        self.create_record({'flavor': 'watermelon', 'author': 'Ethan'})
+        sorting = [Sort('author', 1)]
+        records, _ = self.storage.get_all(sorting=sorting, **self.storage_kw)
+        # Some interesting values to compare against
+        VALUES = ['A', 'Z', '', 0, 4]
```
Comment (leplatrem) [inline kinto/core/storage/testing.py]: µnit: I don't think it's necessary to have uppercase here

**Comment by leplatrem** [comment]
> I'm not sure whether this qualifies as a major version and/or API version.

The way I see it:

* You fixed inconsistencies and bugs
* Those fixes deserve a mention in the API changelog, so we should bump its version
* I don't think it requires a major dump for the API version ( #1252 didn't :) )
* It doesn't break anything on the Python side, so major dump in the package version either

**Comment by glasserc** [comment]
I only have two concerns about HTTP API compatibility:

- We break the technique shown in https://github.com/Kinto/kinto/issues/344 (but that was never documented, so I'm not super worried about it)
- We implicitly discourage the use of fields with the name `has_foo`

**Comment by leplatrem** [comment]
> We break the technique shown in #344 

With all due respect, it was very hacky.

> We implicitly discourage the use of fields with the name has_foo

Well, I agree, but acceptable IMO since that's also true for other fields as shown in #1004

**Comment by Natim** [comment]
This is awesome thank you for tackling it!

**Pull request merged**